src/Elastic.Transport/Components/Pipeline/DefaultResponseBuilder.cs

```diff
@@ -224,68 +224,69 @@ private async ValueTask<TResponse> SetBodyCoreAsync<TResponse>(bool isAsync,
 
+			return response;
+		}
+
+		// TODO: Handle empty data in a nicer way as throwing exceptions has a cost we'd like to avoid!
+		// ie. check content-length (add to ApiCallDetails)? Content-length cannot be retrieved from a GZip content stream which is annoying.
+		try
+		{
+			if (requiresErrorDeserialization && TryGetError(details, requestData, responseStream, out var error) && error.HasError())
+			{
+				response = new TResponse();
+				SetErrorOnResponse(response, error);
 				return response;
 			}
 
-			// TODO: Handle empty data in a nicer way as throwing exceptions has a cost we'd like to avoid!
-			// ie. check content-length (add to ApiCallDetails)? Content-length cannot be retrieved from a GZip content stream which is annoying.
-			try
-			{
-				if (requiresErrorDeserialization && TryGetError(details, requestData, responseStream, out var error) && error.HasError())
-				{
-					response = new TResponse();
-					SetErrorOnResponse(response, error);
-					return response;
-				}
+			if (!requestData.ValidateResponseContentType(mimeType))
+				return default;
 
-				if (!requestData.ValidateResponseContentType(mimeType))
-					return default;
+			var beforeTicks = Stopwatch.GetTimestamp();
 
-				var beforeTicks = Stopwatch.GetTimestamp();
+			if (isAsync)
+				response = await serializer.DeserializeAsync<TResponse>(responseStream, cancellationToken).ConfigureAwait(false);
+			else
+				response = serializer.Deserialize<TResponse>(responseStream);
 
-				if (isAsync)
-					response = await serializer.DeserializeAsync<TResponse>(responseStream, cancellationToken).ConfigureAwait(false);
-				else
-					response = serializer.Deserialize<TResponse>(responseStream);
+			var deserializeResponseMs = (Stopwatch.GetTimestamp() - beforeTicks) / (Stopwatch.Frequency / 1000);
 
-				var deserializeResponseMs = (Stopwatch.GetTimestamp() - beforeTicks) / (Stopwatch.Frequency / 1000);
+			if (deserializeResponseMs > OpenTelemetry.MinimumMillisecondsToEmitTimingSpanAttribute && OpenTelemetry.CurrentSpanIsElasticTransportOwnedHasListenersAndAllDataRequested)
+				Activity.Current?.SetTag(OpenTelemetryAttributes.ElasticTransportDeserializeResponseMs, deserializeResponseMs);
 
-				if (deserializeResponseMs > OpenTelemetry.MinimumMillisecondsToEmitTimingSpanAttribute && OpenTelemetry.CurrentSpanIsElasticTransportOwnedHasListenersAndAllDataRequested)
-					Activity.Current?.SetTag(OpenTelemetryAttributes.ElasticTransportDeserializeResponseMs, deserializeResponseMs);
+			if (!response.LeaveOpen)
+				responseStream.Dispose();
 
-				return response;
-			}
-			catch (JsonException ex) when (ex.Message.Contains("The input does not contain any JSON tokens"))
-			{
-				return default;
-			}
+			return response;
+		}
+		catch (JsonException ex) when (ex.Message.Contains("The input does not contain any JSON tokens"))
+		{
+			responseStream.Dispose();
+			return default;
```

Comment:
We should only dispose this stream, if it was internally created by the `DefaultResponseBuilder` (e.g. when `DisableDirectStreaming` is enabled), not if it is the original network stream.

IMO creation and disposal should be as close as possible and since the `HttpRequestInvoker` already handles disposal of the network stream, we should not do it a second time here.

Reply:
Here I was leaning toward the response builder knowing when it was done and to release the resource. It's enabled adding many more test cases. I'll play around some more with only disposing the memory stream from `DisableDirectStreaming` and see about moving my test cases up to the integration layer.

Reply:
Q? why do we need the distinction on DisableDirectStreaming to dispose?

Reply:
Discussed offline